### PR TITLE
Attempt to get Neo4j working with reactive

### PR DESCRIPTION
**Pull request by mraible**
This is an attempt to fix Neo4j so it works with reactive. 

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

Currently fails with the following error:

```
No property exists found for type PersistentAuditEvent!
```

Here's the JDL I'm using to create a Neo4j reactive gateway:

```
application {
  config {
    baseName gateway,
    packageName com.okta.developer.gateway,
    applicationType gateway,
    authenticationType oauth2,
    databaseType neo4j,
    devDatabaseType neo4j,
    prodDatabaseType neo4j,
    enableHibernateCache false,
    reactive true,
    searchEngine elasticsearch,
    serviceDiscoveryType eureka,
    testFrameworks [protractor]
  }
  entities Blog, Post, Tag, Product
}

application {
  config {
    baseName blog,
    packageName com.okta.developer.blog,
    applicationType microservice,
    authenticationType oauth2,
    databaseType couchbase,
    devDatabaseType couchbase,
    prodDatabaseType couchbase,
    enableHibernateCache false,
    reactive true,
    searchEngine elasticsearch,
    serverPort 8081,
    serviceDiscoveryType eureka
  }
  entities Blog, Post, Tag
}

application {
  config {
    baseName store,
    packageName com.okta.developer.store,
    applicationType microservice,
    authenticationType oauth2,
    databaseType mongodb,
    devDatabaseType mongodb,
    prodDatabaseType mongodb,
    enableHibernateCache false,
    reactive true
    searchEngine elasticsearch,
    serverPort 8082,
    serviceDiscoveryType eureka
  }
  entities Product
}

entity Blog {
  name String required minlength(3),
  handle String required minlength(2)
}

entity Post {
  title String required,
  content TextBlob required,
  date Instant required
}

entity Tag {
  name String required minlength(2)
}

entity Product {
  title String required,
  price BigDecimal required min(0),
  image ImageBlob
}

relationship ManyToOne {
  Blog{user(login)} to User,
  Post{blog(name)} to Blog
}

relationship ManyToMany {
  Post{tag(name)} to Tag{post}
}

paginate Post, Tag with infinite-scroll
paginate Product with pagination

microservice Product with store
microservice Blog, Post, Tag with blog

```

### CODE REVIEW

**Comment by mraible** [comment]
Related to this issue, I noticed that `AccountResourceIT.java` has the following unused imports.

```
import com.okta.developer.gateway.service.UserService;
import reactor.core.publisher.Mono;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
```

**Comment by atomfrede** [comment]
Thanks @mraible I will test it out this evening.

**Comment by mraible** [comment]
This could be related to https://github.com/jhipster/generator-jhipster/pull/11612 since it fails to compile with a missing `AuditEventService` too.

**Comment by atomfrede** [comment]
@mraible I tested your changes and adapted small things (also adding `neo4j` to database type check) and it fails. I removed elastic search and it work. It seems like a compatibility problem with sdn/rx and spring data elasticsearch. I will talk to Gerrit and Michael about that.

**Comment by atomfrede** [comment]
@mraible As a workaround you can do the following in `DatabaseConfiguration`:

```java
@EnableReactiveNeo4jRepositories(basePackageClasses = {AuthorityRepository.class})
@EnableReactiveElasticsearchRepositories(basePackageClasses = {UserSearchRepository.class})
```
Instead of just defining the package for the elastic repositories. Important is also to define at least one class from the "normal" repositories.

**Comment by meistermeier** [comment]
First of all: There is a missing bit in SDN/RX that will Spring Data switch to multi data module mode. I have a fix locally but need to tailor it a little bit more to get it into our main development branch.
But also with this in place, shared entities with multiple store specific annotations (`@Node` / `@Document`) like `User` won't work. Even though all repositories are extending the store specific ones (`ReactiveNeo4jRepository`/`ReactiveElasticsearchRepository`) the Spring Data mechanism will also look at the entity annotations later on and then tries to generate a repository of type `ReactiveElasticsearchRepository` out of a `ReactiveNeo4jRepository`.
The solution @atomfrede provides would be the option to choose if you want to keep the annotations  on the shared entities like `User`.

Just for completeness and you may find the right place to address this: At least the `ReactiveNeo4jRepository` extending interfaces are annotated with `@Repository`. This should be avoided because it will add more infrastructure than needed or desired.

**Comment by atomfrede** [comment]
:heart: @meistermeier for the heads up and debugging, really appreciate that! @mraible If you are okay I will implement the workaround (based on your branch).

**Comment by mraible** [comment]
> @mraible If you are okay I will implement the workaround (based on your branch).

Please do! I'll change it to ready for review so you can merge once it's all working.

**Comment by atomfrede** [comment]
@mraible Can I push on your branch/fork?

**Comment by mraible** [comment]
@atomfrede Yep!

**Comment by atomfrede** [comment]
Awesome, will do it later this evening.

**Comment by atomfrede** [comment]
I am too stupid or just confused. How exactly is collaboration across forks expected to work? I checked out everything locally via 
```
git fetch upstream refs/pull/11612/head
git checkout -b mraible/reactive-neo4j FETCH_HEAD
```
I rebased against master and pushed to my origin (which is obviously not correct). Where am I supposed to push to bring the changes to this PR?

**Comment by mraible** [comment]
@atomfrede I believe you need to do:

```
git remote add mraible git@github.com:mraible/generator-jhipster.git
git fetch mraible
git checkout reactive-neo4j
# commit your changes
git push mraible reactive-neo4j
```

**Comment by atomfrede** [comment]
Let me try. I added your fork as another remote, but with https url. It didn't work (no access, while pushing).

**Comment by atomfrede** [comment]
Okay, with the git url it works. Now bringing my changes to your branch

**Comment by atomfrede** [comment]
I have tested the following:

* reactive gateway with neo4j & elasticsearch from your example jdl :heavy_check_mark: 
* reactive monolith with neo4j & no elasticsearch, jwt :heavy_check_mark: (with changes from master, did not rebase)
* reactive monolith with neo4j & elasticsearch, jwt 

We support elastic now also when using reactive. The current prompt setup does not reflect that. I have updated that too.
Two failing tests during verify/check due to blockhound, I think we should look at that separately from this PR.

File: generators/server/templates/src/main/java/package/repository/search/UserSearchRepository.java.ejs [hunk 1]
```diff
@@ -40,7 +40,7 @@ import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
 /**
  * Spring Data Elasticsearch repository for the User entity.
  */
-public interface UserSearchRepository extends <% if (reactive) {%>Reactive<% } %>ElasticsearchRepository<<%= asEntity('User') %>, <% if (databaseType === 'sql' && authenticationType !== 'oauth2') { %>Long<% } %><% if (databaseType === 'cassandra' || databaseType === 'mongodb' || authenticationType === 'oauth2') { %>String<% } %>><% if (reactive) {%>, UserSearchRepositoryInternal<% } %> {
+public interface UserSearchRepository extends <% if (reactive) {%>Reactive<% } %>ElasticsearchRepository<<%= asEntity('User') %>, <% if (databaseType === 'sql' && authenticationType !== 'oauth2') { %>Long<% } %><% if (databaseType === 'cassandra' || databaseType === 'mongodb' || databaseType === 'neo4j' || authenticationType === 'oauth2') { %>String<% } %>><% if (reactive) {%>, UserSearchRepositoryInternal<% } %> {
```
Comment (murdos) [inline generators/server/templates/src/main/java/package/repository/search/UserSearchRepository.java.ejs:43]:
```suggestion
public interface UserSearchRepository extends <% if (reactive) {%>Reactive<% } %>ElasticsearchRepository<<%= asEntity('User') %>, <% if (databaseType === 'sql' && authenticationType !== 'oauth2') { %>Long<% } %><% if (['cassandra', 'mongodb', 'neo4j'].includes(databaseType) || authenticationType === 'oauth2') { %>String<% } %>><% if (reactive) {%>, UserSearchRepositoryInternal<% } %> {
```

Comment (atomfrede) [reply generators/server/templates/src/main/java/package/repository/search/UserSearchRepository.java.ejs:43]: Thanks

File: generators/server/templates/src/main/java/package/config/DatabaseConfiguration.java.ejs [hunk 2]
```diff
@@ -151,8 +156,11 @@ import java.util.UUID;
     <%_ } _%>
 @EnableTransactionManagement
 <%_ } _%>
+<%_ if (databaseType === 'neo4j') { _%>
+@Enable<% if (reactive) { %>Reactive<% } %>Neo4jRepositories(basePackageClasses = {RepositoryMarker.class})    
+<%_ } _%>   
 <%_ if (searchEngine === 'elasticsearch') { _%>
-@Enable<% if (reactive) { %>Reactive<% } %>ElasticsearchRepositories("<%= packageName %>.repository.search")
+@Enable<% if (reactive) { %>Reactive<% } %>ElasticsearchRepositories(basePackageClasses = {SearchRepositoryMarker.class})
```
Comment (murdos) [inline generators/server/templates/src/main/java/package/config/DatabaseConfiguration.java.ejs:163]: I wonder if the same workaround that is used for mongodb and couchbase would not work. Has it been tried?

Comment (atomfrede) [reply generators/server/templates/src/main/java/package/config/DatabaseConfiguration.java.ejs:163]: You mean with the filters? I didn't try but maybe it works too, worth a try.

Comment (murdos) [reply generators/server/templates/src/main/java/package/config/DatabaseConfiguration.java.ejs:163]: Yes, with the filters, sorry for not having been explicit...

Comment (atomfrede) [reply generators/server/templates/src/main/java/package/config/DatabaseConfiguration.java.ejs]: :heavy_check_mark: Done

File: generators/server/templates/src/main/java/package/repository/UserRepository.java.ejs [hunk 3]
```diff
@@ -121,7 +121,6 @@ import static <%= packageName %>.config.Constants.ID_DELIMITER;
     }
 _%>
 <%_ if ((databaseType === 'sql' && !reactive) || databaseType === 'mongodb' || databaseType === 'neo4j' || databaseType === 'couchbase') { _%>
-@Repository
```
Comment (murdos) [inline generators/server/templates/src/main/java/package/repository/UserRepository.java.ejs:124]: I don't think it should be removed ^^

Comment (murdos) [reply generators/server/templates/src/main/java/package/repository/UserRepository.java.ejs:124]: > 
> 
> I don't think it should be removed ^^

Excepted this last detail, LGTM!

Comment (atomfrede) [reply generators/server/templates/src/main/java/package/repository/UserRepository.java.ejs:124]: Maybe you are right. For sql we don't use a specific base repository. Will add it again.

**Comment by atomfrede** [comment]
Thanks @murdos for having a look too!

**Pull request merged**